### PR TITLE
nixos/meilisearch: init

### PR DESCRIPTION
**Pull request by Br1ght0ne**
###### Motivation for this change
https://github.com/NixOS/nixpkgs/pull/84574

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [X] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by filalex77** [comment]
@GrahamcOfBorg test meilisearch

**Comment by filalex77** [comment]
@GrahamcOfBorg test meilisearch

**Comment by filalex77** [comment]
@GrahamcOfBorg test meilisearch

File: nixos/modules/misc/ids.nix [hunk 1]
```diff
@@ -346,6 +346,7 @@ in
       zoneminder = 314;
       paperless = 315;
       #mailman = 316;  # removed 2019-08-30
+      meilisearch = 317;
```
Comment (aanderse) [inline nixos/modules/misc/ids.nix:349]: https://github.com/NixOS/rfcs/blob/master/rfcs/0052-dynamic-ids.md

File: nixos/modules/services/search/meilisearch.nix [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+in {
+
+  meta.maintainers = with maintainers; [ filalex77 ];
+
+  ###### interface
+
+  options.services.meilisearch = {
+    enable = mkEnableOption "MeiliSearch - a RESTful search API";
+
+    package = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:16]: Unless there are multiple options or variants of the `melisearch` package there isn't any real point to this.

File: nixos/modules/services/search/meilisearch.nix [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      };
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/meilisearch";
+        User = "meilisearch";
+      };
+    };
+
+    environment.systemPackages = [ cfg.package ];
+
+    users = {
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:88]: You might consider simplifying and using `DynamicUser` for this.

File: nixos/modules/services/search/meilisearch.nix [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      groups.meilisearch.gid = config.ids.gids.meilisearch;
+      users.meilisearch = {
+        uid = config.ids.uids.meilisearch;
+        description = "MeiliSearch daemon user";
+        home = cfg.dbPath;
+        createHome = true;
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:94]: Use either `serviceConfig.StateDirectory` or `systemd.tmpfiles.rules` to provision these types of directories.

File: nixos/modules/services/search/meilisearch.nix [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      environment = {
+        MEILI_DB_PATH = cfg.dbPath;
+        MEILI_HTTP_ADDR = "${cfg.listenAddress}:${toString cfg.listenPort}";
+        MEILI_MASTER_KEY = mkIf (cfg.masterKey != null) (toString cfg.masterKey);
+        MEILI_NO_ANALYTICS = toString cfg.noAnalytics;
+        MEILI_ENV = cfg.environment;
+      };
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/meilisearch";
+        User = "meilisearch";
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:82]: You didn't specify `Group`.

File: nixos/modules/services/search/meilisearch.nix [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+      type = types.path;
+    };
+
+    environment = mkOption {
+      description = "Defines the running environment of MeiliSearch.";
+      default = "development";
+      type = types.enum [ "development" "production" ];
+    };
+
+    masterKey = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:47]: This isn't a secret key, right?

File: nixos/modules/services/search/meilisearch.nix [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+      type = types.str;
+    };
+
+    listenPort = mkOption {
+      description = "MeiliSearch port to listen on.";
+      default = 7700;
+      type = types.port;
+    };
+
+    dbPath = mkOption {
```
Comment (aanderse) [inline nixos/modules/services/search/meilisearch.nix:35]: If this is just internal application state I'd highly suggest using `DynamicUser` and then removing this option.

**Review by aanderse — CHANGES_REQUESTED** [review]

**Comment by happysalada** [comment]
@aanderse I'm taking this over. This is 90% complete though. I was thinking rather than just copying and pasting, that we could merge this and I could add the finishing touches on top. (this would keep attribution).

Just to detail what I think is still missing in this PR
- it needs an assertion that when release mode is production, you need to have a master key
- masterKey is indeed a secret. I'm pondering the best way to handle this. I've open an issue upstream for them to allow secrets to be read from files https://github.com/meilisearch/MeiliSearch/issues/1668
- I'm not sure we need `environment.systemPackages` to include meilisearch
- There are probably 2-3 environment variables missing https://docs.meilisearch.com/reference/features/configuration.html#dumps-destination

Responding to a few comments
- I think it makes sense to keep the package option here, since the package can be compiled with a different feature set

**Comment by mohe2015** [comment]
Wouldn't it work to just base your PR on top of this one and then squash merge. Or amend to the commit and add yourself as co-author (or the other way around)?

**Comment by happysalada** [comment]
sure, let me try to make a PR based off those 2 commits.

**Comment by Br1ght0ne** [comment]
@happysalada Let me know if I can help in any way. I'm also fine with closing this PR in favor of yours.

**Comment by happysalada** [comment]
don't worry about it, I can cherry pick those commits, not a problem at all.
Thanks again for making those in the first place !

**Comment by happysalada** [comment]
Just opened a PR here https://github.com/NixOS/nixpkgs/pull/137078

**Pull request closed without merging**